### PR TITLE
fix[tool]: include structs in `-f interface` output

### PR DESCRIPTION
**Pull request by sandbubbles**
### What I did
fix https://github.com/vyperlang/vyper/issues/4282. 

### Commit message

```
the `-f interface` output was missing struct definitions; this commit
fixes the issue.

this commit also does some small code touchups in the
`build_interface_output()` function, and makes the number of newlines
between functions, events and structs consistent - to be 2 `\n`
characters.
```

### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

**Review by charles-cooper — APPROVED** [review]

**Review by charles-cooper — APPROVED** [review]

File: vyper/compiler/output.py [hunk 1]
```diff
@@ -134,6 +134,13 @@ def build_interface_output(compiler_data: CompilerData) -> str:
             encoded_args = "\n    ".join(f"{name}: {typ}" for name, typ in event.arguments.items())
             out = f"{out}event {event.name}:\n    {encoded_args if event.arguments else 'pass'}\n"
 
+    if interface.structs:
```
Comment (charles-cooper) [inline vyper/compiler/output.py:137]: prefer `if len(interface.structs) > 0:`

i see that's how it's done in the rest of the function, i'll push a fix

File: vyper/compiler/output.py [hunk 2]
```diff
@@ -134,6 +134,13 @@ def build_interface_output(compiler_data: CompilerData) -> str:
...
+        out = f"{out}\n# Structs\n\n"
+        for struct in interface.structs.values():
+            out = f"{out}struct {struct.name}:\n"
+            for member_name, member_type in struct.members.items():
+                out = f"{out}\t{member_name}: {member_type}\n"
+
```
Comment (charles-cooper) [inline vyper/compiler/output.py:143]: some more newlines would be nice

Comment (charles-cooper) [reply vyper/compiler/output.py:138]: b2a64a2da84de66442a20400ce7701c1c4ac792d

File: vyper/compiler/output.py (hunk 1 above)
Comment (charles-cooper) [reply vyper/compiler/output.py]: b2a64a2da84de66442a20400ce7701c1c4ac792d

File: vyper/compiler/output.py [hunk 3]
```diff
@@ -134,6 +134,13 @@ def build_interface_output(compiler_data: CompilerData) -> str:
             encoded_args = "\n    ".join(f"{name}: {typ}" for name, typ in event.arguments.items())
             out = f"{out}event {event.name}:\n    {encoded_args if event.arguments else 'pass'}\n"
 
+    if interface.structs:
+        out = f"{out}\n# Structs\n\n"
+        for struct in interface.structs.values():
+            out = f"{out}struct {struct.name}:\n"
+            for member_name, member_type in struct.members.items():
+                out = f"{out}\t{member_name}: {member_type}\n"
```
[the remark this answers is not in the document]
Comment (charles-cooper) [reply vyper/compiler/output.py]: b2a64a2da84de66442a20400ce7701c1c4ac792d

**Comment by pcaversaccio** [comment]
I quickly teste the PR, and I think it's incomplete. It does not list the imported (via module) structs. Example from 🐍 snekmate:

```console
vyper -f interface src/snekmate/utils/mocks/batch_distributor_mock.vy
```

outputs

```console
# Functions

@payable
@external
def distribute_ether(data: Batch):
    ...


@external
def distribute_token(token: .../snekmate/venv/Lib/site-packages/vyper/builtins/interfaces/IERC20.vyi, data: Batch):
    ...
```

The structs:

```console
# Structs

struct Transaction:
    recipient: address
    amount: uint256


struct Batch:
    txns: DynArray[Transaction, 255]
```

are missing in the interface for `batch_distributor_mock.vy`.

Furthermore (not related to this PR), I would like to highlight that if the usage of interfaces as types displayed within `-f interface` outputs is weird (see above `def distribute_token(token: .../snekmate/venv/Lib/site-packages/vyper/builtins/interfaces/IERC20.vyi, data: Batch):`.

**Comment by cyberthirst** [comment]
for output structs we output the names of their fields although they are tuples in the abi, is this intended?
```vyper
# test.vyi
struct Foo:
    x: uint256

def foo() -> Foo:
    ...
 ```

```
-f abi (see "name": "x",)

```json
[
  {
    "stateMutability": "nonpayable",
    "type": "function",
    "name": "foo",
    "inputs": [],
    "outputs": [
      {
        "name": "",
        "type": "tuple",
        "components": [
          {
            "name": "x",
            "type": "uint256"
          }
        ]
      }
    ]
  }

```

**Comment by cyberthirst** [comment]
bit unrelated, but if an interface `implements` other interface and doesn't implement all the methods, we get the following exception:
```
vyper.exceptions.InterfaceViolation: Contract does not implement all interface functions: foobar()
```
`Contract` does not implement.. although it can be triggered via eg `vyper -f abi i.vyi`

**Pull request merged**